File: python/linux_drivers/ethernet/test-E-TC32.py
#! /usr/bin/python3
#
# Copyright (c) 2018 Warren J. Jasper <wjasper@ncsu.edu>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

from E_TC32 import *
import time

def main():
  if len(sys.argv) == 2:
    device = []
    device.append(mccEthernetDevice(ETC32_PID,sys.argv[1]))  
  else:
    # Find a E-TC32 device on the subnet
    device = mccDiscover(ETC32_PID)

  print('Number of devices found = ', len(device))
  if (len(device) > 0):
    for i in range(len(device)):
      device[i].printDeviceInfo()
  else:
    exit(0)

  # Open the device
  device[0].mccOpenDevice()

  # initalize the class
  etc32 = E_TC32(device[0])

  etc32.units = CELSIUS
  etc32.wait = 0x0

  while(True):
    print("\nE-TC32 Testing")
    print("----------------")
    print("Hit 'a' to print alarm status")
    print("Hit 'b' to blink")
    print("Hit 'C' for A/D system calibration.")
    print("Hit 'd' to test digitial IO.")
    print("Hit 'e' to exit")
    print("Hit 'j' for CJC compensation")
    print("Hit 'r' to reset the device")
    print("Hit 'n' to get networking information")
    print("Hit 's' to get Status")
    print("hit 'R' to read System Memory Map")
    print("Hit 't' for temperature.")
    print("Hit 'T' for multiple temperature readings.")
    print("Hit 'v' for version and calibration date.")

    ch = input('')
    if ch == 'b':
      count = input('Enter number of times to blink: ')
      etc32.Blink(int(count))
    elif ch == 'a':
      print('Alarm Status:',etc32.alarm_status)
      etc32.ClearAlarmStatus(0xffffffff)
      print('Alarm Configuration:', etc32.alarm_config)
      print('Alarm threshold1:', etc32.alarm_threshold1)
      print('Alarm threshold2:', etc32.alarm_threshold2)
    elif ch == 'd':
      print("\n Testing Digital I/O ...")
      value = int(input('Enter a number [0-0xf]: '),16)
      etc32.DOut(value)
      value_base = etc32.DIn()
      value_latch = etc32.DOut_R()
      print('The number you entered = ',hex(value_base), '  Latched value = ', hex(value_latch))
    elif ch == 'C':
      print('Performing A/D system offset calibration')
      etc32.ADCal()
      if(etc32.Status == 1):  # EXP detected
        (date_base, date_exp) = etc32.FactoryCalDate_R()
        print('Factory Calibration Date (base): ', date_base)
        print('Factory Calibration Date (exp): ', date_exp)
        (date_base, date_exp) = etc32.FieldCalDate_R()
        print('Field Calibration Date (base): ', date_base)
        print('Field Calibration Date (exp): ', date_exp)
      else:
        date_base = etc32.FactoryCalDate_R()
        print('Factory Calibration Date (base): ', date_base)
        date_base = etc32.FieldCalDate_R()
        print('Field Calibration Date (base): ', date_base)
      print('')
      etc32.FactoryCoef_R()
      print('Printing Factory Calibration coefficients for base')
      print('ADC 0: slope_60_base: ',format(etc32.calCoefFactory[0].slope_60_base,'.6f'))
      print('ADC 0: slope_50_base: ',format(etc32.calCoefFactory[0].slope_50_base,'.6f'))
      print('ADC 0: intercept_60_base: ',format(etc32.calCoefFactory[0].intercept_60_base,'.6f'))
      print('ADC 0: intercept_50_base: ',format(etc32.calCoefFactory[0].intercept_50_base,'.6f'))
      print('ADC 1: slope_60_base: ',format(etc32.calCoefFactory[1].slope_60_base,'.6f'))
      print('ADC 1: slope_50_base: ',format(etc32.calCoefFactory[1].slope_50_base,'.6f'))
      print('ADC 1: intercept_60_base: ',format(etc32.calCoefFactory[1].intercept_60_base,'.6f'))
      print('ADC 1: intercept_50_base: ',format(etc32.calCoefFactory[1].intercept_50_base,'.6f'))
      print('')
      etc32.FieldCoef_R()
      print('Printing Field Calibration coefficients for base')
      print('ADC 0: slope_60_base: ',format(etc32.calCoefField[0].slope_60_base,'.6f'))
      print('ADC 0: slope_50_base: ',format(etc32.calCoefField[0].slope_50_base,'.6f'))
      print('ADC 0: intercept_60_base: ',format(etc32.calCoefField[0].intercept_60_base,'.6f'))
      print('ADC 0: intercept_50_base: ',format(etc32.calCoefField[0].intercept_50_base,'.6f'))
      print('ADC 1: slope_60__base: ',format(etc32.calCoefField[1].slope_60_base,'.6f'))
      print('ADC 1: slope_50__base: ',format(etc32.calCoefField[1].slope_50_base,'.6f'))
      print('ADC 1: intercept_60_base: ',format(etc32.calCoefField[1].intercept_60_base,'.6f'))
      print('ADC 1: intercept_50_base: ',format(etc32.calCoefField[1].intercept_50_base,'.6f'))
    elif ch == 'j':
      for i in range(32):
        temperature = etc32.CJC(i)
        print('Channel =',i,'  Temperature = ',format(temperature,'.2f'),'C   ', format(temperature*9./5.+32., '.2f'),'F')
      print('\nReading Multiple CJC sensor values:')
      value = etc32.CJCMultiple(0xffffffff)
      for i in range(32):
        print('Channel =',i,'  Temperature = ',format(value[i],'.2f'),'C   ', format(value[i]*9./5.+32., '.2f'),'F')
    elif ch == 't':
      chan = int(input('Enter channel number [0-31]: '))
      ttype = input('Enter thermocouple type [J,K,R,S,T,N,E,B]: ')
      if (ttype == 'j' or ttype == 'J'):
        tc_type = TC_TYPE_J
      elif (ttype == 'k' or ttype == 'K'):
        tc_type = TC_TYPE_K
      elif (ttype == 'r' or ttype == 'R'):
        tc_type = TC_TYPE_R
      elif (ttype == 's' or ttype == 'S'):
        tc_type = TC_TYPE_S
      elif (ttype == 't' or ttype == 'T'):
        tc_type = TC_TYPE_T
      elif (ttype == 'n' or ttype == 'N'):
        tc_type = TC_TYPE_N
      elif (ttype == 'e' or ttype == 'E'):
        tc_type = TC_TYPE_E
      elif (ttype == 'b' or ttype == 'B'):
        tc_type = TC_TYPE_B
      else:
        print('Unknown thermocouple type.')
        break
      etc32.TinConfig_R()
      for i in range(64):
        etc32.config_values[i] = CHAN_DISABLE
      etc32.config_values[chan] = tc_type
      etc32.TinConfig_W()
      for i in range(20):
        temperature = etc32.Tin(chan, CELSIUS, 1)
        print('Channel =',chan,'  Temperature = ',format(temperature,'.2f'),'C   ',format(temperature*9./5.+32.,'.2f'),'F')
        time.sleep(1)
    elif ch == 'e':
      etc32.device.sock.close()
      exit(0)
    elif ch == 'r':  
      etc32.Reset()
    elif ch == 's':
      # Print the calibration data and status
      if (etc32.status == 0):
        print('No EXP detected.')
      else:
        print('EXP detected.')
      etc32.TinStatus()
      print('TinStatus = ',hex(etc32.Tin_status[0]))
      etc32.OTDStatus()
      print('OTDStatus = ',hex(etc32.OTD_status[0]))
    elif ch == 'T':
      print('Read Multiple Thermocouple channels')
      ttype = input('Enter thermocouple type [J,K,R,S,T,N,E,B]: ')
      if (ttype == 'j' or ttype == 'J'):
        tc_type = TC_TYPE_J
      elif (ttype == 'k' or ttype == 'K'):
        tc_type = TC_TYPE_K
      elif (ttype == 'r' or ttype == 'R'):
        tc_type = TC_TYPE_R
      elif (ttype == 's' or ttype == 'S'):
        tc_type = TC_TYPE_S
      elif (ttype == 't' or ttype == 'T'):
        tc_type = TC_TYPE_T
      elif (ttype == 'n' or ttype == 'N'):
        tc_type = TC_TYPE_N
      elif (ttype == 'e' or ttype == 'E'):
        tc_type = TC_TYPE_E
      elif (ttype == 'b' or ttype == 'B'):
        tc_type = TC_TYPE_B
      else:
        print('Unknown thermocouple type.')
        break
      for i in range(31):
        etc32.config_values[i] = tc_type
      etc32.TinConfig_W()
      # just read the even channels
      etc32.channel_mask[0] = 0x55555555
      wait = 1
      units = CELSIUS
      value = etc32.TinMultiple(wait, units, 0x55555555)
      for i in range(16):
        print('Channel ',2*i,'  Temperature = ', format(value[i],'.2f'))
      # turn off the LED on the fron pannel
      etc32.config_measure[0] = 1   # disable the OTC which turn off the LED
      etc32.MeasureConfig_W()
      etc32.config_measure[0] = 0   # enable the OTD
      etc32.MeasureConfig_W()
    elif ch == 'n':
      value = etc32.NetworkConfig()
      print('Network configuration values: ')
      print('  IP address = ', value[0])
      print('  subnet mask = ', value[1])
      print('  gateway = ', value[2])
      print('  MAC:', \
            hex(etc32.device.MAC>>40&0xff)[2:].zfill(2)+':'+\
            hex(etc32.device.MAC>>32&0xff)[2:].zfill(2)+':'+\
            hex(etc32.device.MAC>>24&0xff)[2:].zfill(2)+':'+\
            hex(etc32.device.MAC>>16&0xff)[2:].zfill(2)+':'+\
            hex(etc32.device.MAC>>8&0xff)[2:].zfill(2)+':'+\
          hex(etc32.device.MAC&0xff)[2:].zfill(2))
    elif ch == 'v':
      etc32.Version()
      print('Communications micro firmware version:', str(etc32.version[0]>>8&0xff)+'.'+str(etc32.version[0]&0xff))
      print('Communications micro bootloader firmware version:', str(etc32.version[1]>>8&0xff)+'.'+str(etc32.version[1]&0xff))
      print('Base measurement micro firmware version:', str(etc32.version[2]>>8&0xff)+'.'+str(etc32.version[2]&0xff))
      print('Base measurement micro bootloader firmware version:', str(etc32.version[3]>>8&0xff)+'.'+str(etc32.version[3]&0xff))
      print('EXP measurement micro firmware version:', str(etc32.version[4]>>8&0xff)+'.'+str(etc32.version[4]&0xff))
      print('EXP measurement micro bootloader firmware version:', str(etc32.version[5]>>8&0xff)+'.'+str(etc32.version[5]&0xff))
    elif ch == 'R':
      print('Reading Settings Memory.')
      for i in range(0x1f):
        value = etc32.SettingsMemory_R(i, 1)
        print('address: ',hex(i), '    value =', value)
    else:
      pass

if __name__ == "__main__":
  main()

File: c/Ethernet/test-E-TC32.c
/*
 *  Copyright (c) 2016 Warren J. Jasper <wjasper@ncsu.edu>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdbool.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include "E-TC32.h"

#include <fcntl.h>
#include <math.h>
#include <ctype.h>

/* Test Program */
int toContinue()
{
  int answer;
  answer = 0; //answer = getchar();
  printf("Continue [yY]? ");
  while((answer = getchar()) == '\0' ||
    answer == '\n');
  return ( answer == 'y' || answer == 'Y');
}

int main(int argc, char**argv)
{
  struct version_t version;
  struct tm date_base, date_exp;
  struct networkDeviceInfo_t network;
  struct in_addr ip_addr;

  DeviceInfo_TC32 device_info;
  uint8_t dioIn[2];
  uint32_t dioOut;
  int i;
  float temperature, CJC_temp;
  uint8_t channel;
  uint8_t tc_type;
  int ch;
  uint8_t options;
  uint8_t buf[32];
  in_addr_t address;

 start:
  device_info.device.connectCode = 0x0;   // default connect code
  device_info.device.frameID = 0;         // zero out the frameID

  if (argc == 2) {
    printf("E-TC32 IP address = %s\n", argv[1]);
    device_info.device.Address.sin_family = AF_INET;
    device_info.device.Address.sin_port = htons(COMMAND_PORT);
    device_info.device.Address.sin_addr.s_addr = INADDR_ANY;
    if (inet_aton(argv[1], &device_info.device.Address.sin_addr) == 0) {
      printf("Improper destination address.\n");
      return -1;
    }
  } else if (discoverDevice(&device_info.device, ETC32_PID) <= 0) {
    printf("No device found.\n");
    return -1;
  }

  if ((device_info.device.sock = openDevice(inet_addr(inet_ntoa(device_info.device.Address.sin_addr)),
					    device_info.device.connectCode)) < 0) {
    printf("Error opening socket\n");
    return -1;
  }

  device_info.units = UNITS_CELSIUS;
  device_info.wait = 0x0;
  for (i = 0; i< 64; i++) {
    device_info.config_values[i] = 0x0;   // disable all channels
  }

  while(1) {
    printf("\nE-TC32 Testing\n");
    printf("----------------\n");
    printf("Hit 'b' to blink\n");
    printf("Hit 'C' for A/D system calibration.\n");
    printf("Hit 'd' to test digitial IO.\n");
    printf("Hit 'e' to exit\n");
    printf("Hit 'j' for CJC compensation\n");
    printf("Hit 'r' to reset the device\n");
    printf("Hit 'n' to get networking information\n");
    printf("Hit 's' to get Status\n");
    printf("hit 'R' to read System Memory Map\n");
    printf("hit 'W' to write System Memory Map\n");
    printf("Hit 't' for temperature.\n");
    printf("Hit 'T' for multiple temperature readings.\n");
    printf("Hit 'v' for version and calibration date.\n");

    while((ch = getchar()) == '\0' || ch == '\n');
    switch(ch) {
      case 'b': /* test to see if LED blinks */
        printf("Enter number or times to blink: ");
        scanf("%hhd", &options);
	BlinkLED_E_TC32(&device_info, options);
        break;
      case 'C':
	printf("Performing A/D system offset calibration.\n");
	ADCal_E_TC32(&device_info);
	break;
      case 'd': /* test DIO */
	printf("\nTesting Digital I/O...\n");
	printf("connect pins DIO input [0-7] <--> DIO Output[0-7]\n");
	do {
	  printf("Enter a  number [0-0xff] : " );
	  scanf("%x", &dioOut);
	  DOutW_E_TC32(&device_info, BASE, dioOut);
	  DIn_E_TC32(&device_info, dioIn);
	  printf("The number you entered = %#x   Value read = %#x\n\n", dioOut, dioIn[0]);
	  for (i = 0; i < 8; i++) {
	    printf("Bit %d = %d\n", i, (dioIn[0]>>i) & 0x1);
	  }
        } while (toContinue());
        break;
      case 'j':
        for (channel = 0; channel < 32; channel++) {
	  CJC_E_TC32(&device_info, channel, &CJC_temp);
	  printf("Channel = %d   Temperature = %.2f C  %.2f F\n", channel, CJC_temp, CJC_temp*9./5. + 32.);
	}
        break;
      case 't':
        printf("Enter channel number [0-31]: ");
	scanf("%hhd", &channel);
	printf("Input Thermocouple type [J,K,R,S,T,N,E,B]: ");
	while((ch = getchar()) == '\0' || ch == '\n');
	switch(ch) {
	  case 'J': 
  	  case 'j':
	    tc_type = TC_TYPE_J; break;
  	  case 'K': 
  	  case 'k': 
	    tc_type = TC_TYPE_K; break;
	  case 'R':
	  case 'r':
	    tc_type = TC_TYPE_R; break;
  	  case 'S':
	  case 's':
	    tc_type = TC_TYPE_S; break;
   	  case 'T':
	  case 't':
	    tc_type = TC_TYPE_T; break;
     	  case 'N':
	  case 'n':
	    tc_type = TC_TYPE_N; break;
       	  case 'E':
  	  case 'e':
	    tc_type = TC_TYPE_E; break;
	  case 'B':
 	  case 'b':
	    tc_type = TC_TYPE_B; break;
	  default: tc_type = TC_TYPE_J; break;
	}
	for (i = 0; i < 64; i++) {
	  device_info.config_values[i] = CHAN_DISABLE;
	}
	device_info.config_values[channel] = tc_type;
	TinConfigW_E_TC32(&device_info);
	for (i = 0; i < 20; i++) {
	  Tin_E_TC32(&device_info, channel, UNITS_CELSIUS, 1, &temperature);
	  printf("Channel = %d   Temperature = %.2f C  %.2F\n", channel, temperature, temperature*9./5. + 32.);
	  sleep(1);
	}
        break;
      case 'r': 
	Reset_E_TC32(&device_info);
	sleep(2);
	goto start;
	break;
      case 's':
        Status_E_TC32(&device_info);
	if (device_info.status == 0) {
	  printf("No EXP detected.\n");
	} else {
	  printf("EXP detected.\n");
	}
	TinStatus_E_TC32(&device_info);
	printf("TinStatus = %#x\n", device_info.Tin_status[0]);
	OTDStatus_E_TC32(&device_info);
	printf("OTDStatus = %#x\n", device_info.OTD_status[0]);
	break;
      case 'T':
	printf("Read Multiple Thermocouple channels\n");
	printf("Input Thermocouple type [J,K,R,S,T,N,E,B]: ");
	while((ch = getchar()) == '\0' || ch == '\n');
	switch(ch) {
	  case 'J': 
  	  case 'j':
	    tc_type = TC_TYPE_J; break;
  	  case 'K': 
  	  case 'k': 
	    tc_type = TC_TYPE_K; break;
	  case 'R':
	  case 'r':
	    tc_type = TC_TYPE_R; break;
  	  case 'S':
	  case 's':
	    tc_type = TC_TYPE_S; break;
   	  case 'T':
	  case 't':
	    tc_type = TC_TYPE_T; break;
     	  case 'N':
	  case 'n':
	    tc_type = TC_TYPE_N; break;
       	  case 'E':
  	  case 'e':
	    tc_type = TC_TYPE_E; break;
	  case 'B':
 	  case 'b':
	    tc_type = TC_TYPE_B; break;
	  default: tc_type = TC_TYPE_J; break;
	}
	for (i = 0; i < 31; i++) {
	  device_info.config_values[i] = tc_type;
	}
	TinConfigW_E_TC32(&device_info);
	// Just read the even channels
	device_info.channel_mask[0] = 0x55555555;
	device_info.wait = 1;
	device_info.units = UNITS_CELSIUS;
	TinMultiple_E_TC32(&device_info);
	for (i = 0; i < 16; i++) {
	  printf("Channel[%d] Temperature = %f\n", 2*i, device_info.Tin_values[i]);
	}
	// Turn off the LED on the front pannel.
	device_info.config_measure[0] = 1;  //disable the OTD which turns off the LED.
	MeasureConfigW_E_TC32(&device_info);
	device_info.config_measure[0] = 0;  //enable the OTD;
	MeasureConfigW_E_TC32(&device_info);
        break;
      case 'v':
	Version_E_TC32(&device_info, &version);
	printf("Communications micro firmware version = %#x\n", version.version_comms);
	printf("Communcations mico bootloader firmware version = %#x\n", version.boot_version_comms);
	printf("Base measurement micro firmware version = %#x\n", version.version_base);
	printf("Base measurement micro bootloader firmware version = %#x\n", version.boot_version_base);
	printf("EXP measurement micro firmware version = %#x\n", version.version_exp);
	printf("EXP measurement micro bootloader firmware version = %#x\n\n", version.boot_version_exp);
	CalDateR_E_TC32(&device_info, &date_base, &date_exp);
	printf("Calibration date = %s\n", asctime(&date_base));
	break;
      case 'e':
	close(device_info.device.sock);
	printf("Success!\n");
        return(0);
        break;
      case 'n':
	NetworkConfig_E_TC32(&device_info, &network);
	ip_addr.s_addr = network.ip_address;
	printf("Device IP address:  %s\n", inet_ntoa(ip_addr));
	ip_addr.s_addr = network.subnet_mask;
	printf("Device subnet mask: %s\n", inet_ntoa(ip_addr));
	ip_addr.s_addr =  network.gateway_address;
	printf("Device gateway address: %s\n", inet_ntoa(ip_addr));
	break;
      case 'R':
	printf("Reading Settings Memory.\n");
	for (i = 0x0; i < 0x1f; i++) {
	  SettingsMemoryR_E_TC32(&device_info, i, 1, buf);
	  printf("address: %#x      value: %d\n", i, buf[0]);
	}
	break;
      case 'W':
	printf("Writing Settings Memory.\n");
	printf("Note: The could be dangerous if you add invalid data!!!\n");

	printf("Enter new default IP address (eg. 192.168.0.101): ");
	scanf("%s", buf);
	address = inet_addr((char *) buf);
	SettingsMemoryW_E_TC32(&device_info, 0x2, 4, (uint8_t*) &address);

	printf("Enter new default netmask (eg. 255.255.255.0): ");
	scanf("%s", buf);
	address = inet_addr((char *) buf);
	SettingsMemoryW_E_TC32(&device_info, 0x6, 4, (uint8_t*) &address);

	printf("Enter new default gateway (eg. 192.168.0.1): ");
	scanf("%s", buf);
	address = inet_addr((char *) buf);
	SettingsMemoryW_E_TC32(&device_info, 0xa, 4, (uint8_t*) &address);

	printf("Enter new Network options (0-3): ");
	scanf("%hhd", buf);
	SettingsMemoryW_E_TC32(&device_info, 0x0, 1, buf);

	printf("Reset or powercycle for changes to take effect.\n");
	
        break;

    }
  }
}
